### DISCUSSION TITLE
rexie vs. idb

### DISCUSSION THREAD

**Discussion by flosse** [devashishdxt/rexie#35] [Q&A] [answered]
@devashishdxt you're the maintainer of both `rexie` and `idb` so probably you can answer the question quite easily:
What's the main difference? And which one is more lightweight or closer to the JS API?

**Comment by ShaddyDC** [comment]
I don't know the libraries as well as him, but I can give you a quick overview until he has time to elaborate. 
`idb` is intended to be a light wrapper to build your own abstractions over, so it's very close to the JS API. For even more fine-grained control over callbacks and the like, you can additionally use the internal `idb-sys`. That also means that some of the rough edges of the call-back API are exposed to the user, and [some cases](https://github.com/devashishdxt/idb/issues/5) are a bit more annoying to handle with idiomatic rust. 
`rexie` on the other hand, adds some more abstractions to make things more convenient to handle, at the cost of allowing less control over the gory details.
Lastly, the author seems to be working on a third library for interacting with IndexedDB as well in [`deli`](https://github.com/devashishdxt/deli). It's a wrapper around `idb`, so if you don't like the ergonomics of `rexie`, that may be worth a look too. 
It should be noted that rexie is more popular than idb is more popular than deli. I don't know to what level the author is planning to support each of these and how mature they are. I can say that the author was reasonably quick to respond to me and to address my issues.

**Comment by devashishdxt** [comment, marked as answer]
I'm definitely guilty of creating so many crates for indexed db. But, here is the complete story:

1. I created `rexie` for a small project I was working on which needed to interact with indexed db. The API was quite nice but it still needed too much code to do very simple things and even though most of the things from indexed db are supported, there are few things which are missing. For example, cursors, migrations, etc.
2. I was going through some `gloo` design discussions and realized that they have a very different approach in their crates. They support both, callback based APIs and wrapper around that to use with `async/await`. So, I created `idb-sys` and `idb`.
3. Finally, I wanted a super simple to use derive macro based crate that can directly store Rust structs in indexed db with minimal code. So, I started working on `deli`. Almost all the things are now supported in `deli` except for migrations and updating the values using cursor.

I think, for someone who doesn't want low level control, `deli` is the most simple crate to use to interact with indexed db.

I hope this answers your question.

**Comment by flosse** [comment]
@devashishdxt Thanks for you answer!
